tests/test_httpproxy_asyncio.py

```diff
@@ -0,0 +1,218 @@
+
+    def test_connect_no_credentials(self):
+        """
+        Tests the proxy connection procedure when there is no basic auth.
+        """
+        # make a fake proxy
+        fake_proxy = MockTransport()
+        # make a fake protocol that we fancy using through the proxy
+        fake_protocol = MockProtocol()
+
+        # create a connect proxy client
+        hcp = HttpConnectProtocol(
+            target_hostport=("example.org", 443),
+            proxy_credentials=None,
+            protocol_factory=lambda: fake_protocol,
+            sslcontext=None,
+            loop=None,
+        )
+
+        switch_over_task = asyncio.get_event_loop().create_task(
+            hcp.switch_over_when_ready()
+        )
+
+        # if it's already done, that'd be odd
+        self.assertFalse(switch_over_task.done())
+
+        # connect the proxy client to the proxy
+        fake_proxy.set_protocol(hcp)
+        hcp.connection_made(fake_proxy)
+
+        # check we got sent a reasonable, uh, hardcoded, CONNECT request
+        self.assertEqual(fake_proxy.buffer, b"CONNECT example.org:443 HTTP/1.0\r\n\r\n")
+        # clear
+        fake_proxy.buffer = b""
+
+        # pretend we got a happy response with some dangling bytes from the
+        # target protocol
+        fake_proxy.pretend_to_receive(
+            b"HTTP/1.0 200 Connection Established\r\n\r\n"
+            b"begin beep boop\r\n\r\n~~ :) ~~"
+        )
+
+        # advance reactor because we have to let coroutines be executed
+        self.loop.advance(1.0)
+
+        # *now* we should have switched over!
+        self.assertTrue(switch_over_task.done())
+
+        transport, protocol = switch_over_task.result()
+
+        # check it was our protocol that was returned
+        self.assertEqual(protocol, fake_protocol)
+
+        # check our protocol received exactly the bytes meant for it
+        self.assertEqual(
+            fake_protocol.received_bytes, b"begin beep boop\r\n\r\n~~ :) ~~"
+        )
+
+    def test_connect_correct_credentials(self):
+        """
+        Tests the proxy connection procedure when there is basic auth.
+        """
+        # make a fake proxy
+        fake_proxy = MockTransport()
+        # make a fake protocol that we fancy using through the proxy
+        fake_protocol = MockProtocol()
+
+        # create a connect proxy client
+        hcp = HttpConnectProtocol(
+            target_hostport=("example.org", 443),
+            proxy_credentials=("user", "secret"),
+            protocol_factory=lambda: fake_protocol,
+            sslcontext=None,
+            loop=None,
+        )
+
+        switch_over_task = asyncio.get_event_loop().create_task(
+            hcp.switch_over_when_ready()
+        )
+
+        # if it's already done, that'd be odd
+        self.assertFalse(switch_over_task.done())
+
+        # connect the proxy client to the proxy
+        fake_proxy.set_protocol(hcp)
+        hcp.connection_made(fake_proxy)
```

Comment:
Looks like this can be in `setUp` or a helper method so we don't repeat the code.